packages/server/lib/controllers/auth/tba.ts

```diff
@@ -0,0 +1,234 @@
+        oauth_timestamp: timestamp.toString(),
+        oauth_callback: callbackUrl
+    };
+
+    const concatenatedParams = collectParameters(oauthParams);
+
+    const baseString = `POST&${percentEncode(tokenRequestUrl)}&${percentEncode(concatenatedParams)}`;
+
+    const emptyTokenSecret = '';
+    const hash = crypto
+        .createHmac('sha256', `${percentEncode(config.oauth_client_secret)}&${percentEncode(emptyTokenSecret)}`)
+        .update(baseString)
+        .digest('base64');
+
+    const authHeader =
+        `OAuth oauth_consumer_key="${percentEncode(config.oauth_client_id)}",` +
+        `oauth_nonce="${nonce}",` +
+        `oauth_timestamp="${timestamp}",` +
+        `oauth_signature_method="${signatureMethod}",` +
+        `oauth_callback="${percentEncode(callbackUrl)}",` +
+        `oauth_signature="${percentEncode(hash)}"`;
+
+    const headers = {
+        Authorization: authHeader
+    };
+
+    const response = await axios.post(tokenRequestUrl, null, { headers });
+
+    const { data } = response;
+
+    if (!data) {
+        await logCtx.error('No data returned from token request');
+        await logCtx.failed();
+
+        res.status(400).send({
+            error: { code: 'no_data_returned_from_token_request' }
+        });
+
+        return;
+    }
+
+    const parsedData = querystring.parse(data);
+    const { oauth_token, oauth_token_secret } = parsedData;
+
+    const authorizeTokenUrl = template.authorization_url as string;
+
+    const fullAuthorizeTokenUrl = interpolateStringFromObject(authorizeTokenUrl, {
+        connectionConfig
+    });
+
+    const session: OAuthSession = {
+        providerConfigKey: providerConfigKey,
+        provider: config.provider,
+        connectionId: connectionId,
+        callbackUrl: callbackUrl,
+        authMode: template.auth_mode,
+        codeVerifier: crypto.randomBytes(24).toString('hex'),
+        id: uuid.v1(),
+        connectionConfig: {
+            ...connectionConfig,
+            oauth_token: oauth_token as string,
+            oauth_token_secret: oauth_token_secret as string,
+            consumer_key: config.oauth_client_id,
+            token_id: tokenId,
+            token_secret: tokenSecret
+        },
+        environmentId: environment.id,
+        webSocketClientId: wsClientId,
+        activityLogId: logCtx.id
+    };
+
+    await oAuthSessionService.create(session);
+
+    const redirectUrl = new URL(fullAuthorizeTokenUrl);
+    redirectUrl.searchParams.append('oauth_token', oauth_token as string);
+    redirectUrl.searchParams.append('state', session.id.replace(/-/g, ''));
+
+    await logCtx.info('Redirecting', {
+        authorizationUri: redirectUrl.toString(),
+        providerConfigKey,
+        connectionId,
+        params: {
+            oauth_token: oauth_token as string,
+            state: session.id.replace(/-/g, '')
+        },
+        connectionConfig
+    });
+
+    res.redirect(redirectUrl.toString());
+});
```

Comment:
a lot of the code above can fail and throw. what kind of response message is sent back to the user in this case?